python/cuml/neighbors/kernel_density.py

```diff
@@ -0,0 +1,406 @@
+            metric_arg = list(self.metric_params.values())[0]
+            distances = pairwise_distances(X, self.X_, metric=self.metric,
+                                           metric_arg=metric_arg)
+        else:
+            distances = pairwise_distances(X, self.X_, metric=self.metric)
+
+        distances = cp.asarray(distances)
+
+        h = self.bandwidth
+        if self.kernel in log_probability_kernels_:
+            distances = log_probability_kernels_[self.kernel](distances, h)
+        else:
+            raise ValueError("Unsupported kernel.")
+
+        log_probabilities = cp.zeros(distances.shape[0])
+        if self.sample_weight_ is not None:
+            distances += cp.log(self.sample_weight_)
+
+        logsumexp_kernel.forall(log_probabilities.size)(
+            distances, log_probabilities)
+        # Note that sklearns user guide is wrong
+        # It says the (unnormalised) probability output for
+        #  the kernel density is sum(K(x,h)).
+        # In fact what they implment is (1/n)*sum(K(x,h))
+        # Here we divide by n in normal probability space
+        # Which becomes -log(n) in log probability space
+        sum_weights = (
+            cp.sum(self.sample_weight_)
+            if self.sample_weight_ is not None
+            else distances.shape[1]
+        )
+        log_probabilities -= np.log(sum_weights)
+
+        # norm
+        log_probabilities = norm_log_probabilities(
+            log_probabilities, self.kernel, h, X.shape[1]
+        )
+
+        return log_probabilities
+
+    def score(self, X, y=None):
+        """Compute the total log-likelihood under the model.
+
+        Parameters
+        ----------
+
+        X : array-like of shape (n_samples, n_features)
+            List of n_features-dimensional data points.  Each row
+            corresponds to a single data point.
+        y : None
+            Ignored.
+
+        Returns
+        -------
+
+        logprob : float
+            Total log-likelihood of the data in X. This is normalized to be a
+            probability density, so the value will be low for high-dimensional
+            data.
+        """
+        return cp.sum(self.score_samples(X))
+
+    def sample(self, n_samples=1, random_state=None):
+        """
+        Generate random samples from the model.
+        Currently, this is implemented only for gaussian and tophat kernels,
+        and the Euclidean metric.
+
+        Parameters
+        ----------
+        n_samples : int, default=1
+            Number of samples to generate.
+        random_state : int, cupy RandomState instance or None, default=None
+
+        Returns
+        -------
+        X : cupy array of shape (n_samples, n_features)
+            List of samples.
+        """
+        if not hasattr(self, "X_"):
+            raise NotFittedError()
+
+        if (self.kernel not in ["gaussian", "tophat"]
+                or self.metric != "euclidian"):
+            raise NotImplementedError()
```

Comment:
It would be nice to see a messgae here to let the user know what's not implemented (e.g. sampling w/ the specified kernel vs the `sample()` method)